### ISSUE TITLE
Test interrupt fails intermittently

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#2192]
We're seeing this error pop up from time to time on travis.ci.  cc @jcrist 

```python
________________________________ test_interrupt ________________________________
[gw0] linux -- Python 3.5.3 /home/travis/miniconda/envs/test-environment/bin/python
@pytest.mark.slow
    def test_interrupt():
        try:
            proc = subprocess.Popen([sys.executable, '-c', code],
                                    stdout=subprocess.PIPE,
                                    stderr=subprocess.PIPE)
            # Wait for scheduler to start
            msg = proc.stdout.readline()
            if msg != 'started\n' and proc.poll() is not None:
                assert False, "subprocess failed"
            # Scheduler has started, send an interrupt
            sigint = signal.CTRL_C_EVENT if os.name == 'nt' else signal.SIGINT
            try:
                proc.send_signal(sigint)
                # Wait a bit for it to die
                start = time()
                while time() - start < 2.0:
                    if proc.poll() is not None:
                        break
                    sleep(0.05)
                else:
>                   assert False, "KeyboardInterrupt Failed"
E                   AssertionError: KeyboardInterrupt Failed
E                   assert False
dask/tests/test_threaded.py:155: AssertionError
```

**Comment by mrocklin** [comment]
It looks like the timeout has been previously extended in 305d2f27 .  The problem may be deeper.

**Comment by jcrist** [comment]
I'm unable to reproduce. Looking at the logs, the failure seems to only happen with the increased lag on the osx builds. Perhaps those builds are even slower? Looking at the times from a passing build it looks that way: https://travis-ci.org/dask/dask/builds/220372113 (~14.5 min compared to ~9 min for an equivalent parallel build on linux).

One "solution" might be to just turn off this test for osx like we already do on windows. It passes fine on local machines, just not on the CI.

**Comment by mrocklin** [comment]
I've seen this pop up on linux builds.  Actually I think that the build from which I copied the printout above is on Linux.

**Comment by jakirkham** [comment]
Seeing this issue again. Please take a look at this [build]( https://travis-ci.org/dask/dask/jobs/225782998#L3592 ) for an example.

**Comment by jakirkham** [comment]
Could we please reopen this issue since it appears to be reoccurring?

**Comment by jakirkham** [comment]
Thanks.

**Comment by jakirkham** [comment]
I think I understand the purpose of this test and it does seem pretty important. Am trying to think of a way to make it more robust. Though must admit this is pretty tricky. Maybe doing this in the same process might be better. 

One way is to change the signal handler in the current process temporarily. CPython does [this]( https://github.com/python/cpython/blob/f0c819acd0f85eafe12a7ff706650cb39d3fbf34/Lib/unittest/test/test_break.py#L29-L40 ) in their test suite by [changing]( https://github.com/python/cpython/blob/f0c819acd0f85eafe12a7ff706650cb39d3fbf34/Lib/unittest/signals.py#L49-L67 ) the signal handler. IMHO this seems like a pretty good option and most similar to what is happening here.

Appears that pytest has some [ways]( https://github.com/pytest-dev/pytest/blob/314d4afa57e1d116010e53c0bceb9f7c608fb666/testing/test_terminal.py#L192-L204 ) of running Python files as well and capturing test failures like this. Though I don't know if they have a way of injecting this signal as well. Might be worth asking them if they have better ideas.

**Comment by jakirkham** [comment]
I loath to mention this given how much work was put into trying to fix this last time. It seems this failure is intermittently showing up again on one of the Python 2.7 builds (oddly always the same one). 😞 Not sure what should be done about. 😕

ref: https://travis-ci.org/dask/dask/jobs/244809778#L6855
ref: https://travis-ci.org/dask/dask/jobs/244927342#L6847
ref: https://travis-ci.org/dask/dask/jobs/242406867#L6735
ref: https://travis-ci.org/dask/dask/jobs/241725181#L6737

**Comment by jcrist** [comment]
Yeah, I saw that. It's been passing for almost 2 months without failure, and then started failing again recently. Diffing the installed dependencies doesn't turn up anything, and the code this is testing hasn't changed recently.

I'm starting to wonder if this is an issue with pytest-xdist and what tests run on what worker. I have little motivation to solve this currently though, mostly frustrated with testing this. Tempted to xfail for 2.7 and reopen this issue to be solved later.

**Comment by jakirkham** [comment]
SGTM.

**Comment by jcrist** [comment]
See #2483.
